llvm/lib/CAS/HashMappedTrie.cpp

```diff
@@ -155,30 +156,23 @@ TrieSubtrie *TrieSubtrie::sink(
 }
 
 struct ThreadSafeHashMappedTrieBase::ImplType {
-  static std::unique_ptr<ImplType> create(size_t StartBit, size_t NumBits) {
+  static ImplType *create(size_t StartBit, size_t NumBits) {
     size_t Size = sizeof(ImplType) + getTrieTailSize(StartBit, NumBits);
     void *Memory = ::malloc(Size);
-    ImplType *Impl = ::new (Memory) ImplType(StartBit, NumBits);
-    return std::unique_ptr<ImplType>(Impl);
+    return ::new (Memory) ImplType(StartBit, NumBits);
   }
 
   TrieSubtrie *save(std::unique_ptr<TrieSubtrie> S) {
-    TrieSubtrie *Ptr = S.get();
+    TrieSubtrie *Ptr = S.release();
     TrieSubtrie *CurrentHead = nullptr;
 
     // Add ownership of "S" to front of the list, so that Root -> S ->
     // Root.Next. Be careful to handle concurrent calls.
-    //
-    // Note: The body of while loop temporarily adds an extra owning reference
-    // to "CurrentHead", releasing the extra reference from the previous
-    // iteration.
-    while (!Root.Next.compare_exchange_weak(CurrentHead, S))
-      S->Next.exchange(std::unique_ptr<TrieSubtrie>(CurrentHead)).release();
-
-    // Drop the extra ownership reference to the previous head of the list.
-    assert(CurrentHead == S.get() &&
+    while (!Root.Next.compare_exchange_weak(CurrentHead, Ptr))
+      Ptr->Next.exchange(CurrentHead);
```

Comment:
I think a longer comment would be helpful, given how  much trouble you and I had figure out how this works (it was more complicated with the `unique_ptr<>` because we were worried about leaking, but still). I'd suggest adding a new paragraph to the comment above, something like:
```
// This works by repeatedly setting S->Next to a candidate value of Root.Next
// (initially nullptr), and then setting Root.Next to S once the candidate matches
// reality.
```